### PR TITLE
stmtctx: move time zone info in stmtctx.StatementContext into TypeCtx

### LINKED ISSUE #47590: move time zone info in `stmtctx.StatementContext` into `TypeCtx`
see description in #47592

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #47590

### What is changed and how it works?

1. Use `StatementContext.TypeCtx.Location` as the timezone in `StatementContext` and remove the dup field `StatementContext.TimeZone`
2. Provide `StatementContext.TimeZone()` to return the inner  location info.
3. Provide some construct methods `NewStmtCtx`, `NewStmtCtxWithTimeZone`, and `Reset` to create or reset `StatementContext` to make sure the inner `TypeCtx` will be properly inited.
4. Replace all usage like `&stmtctx.StatementContext{...}` or `new(stmtctx.StatementContext)` with these constructors. 
5. Provide `SetTimeZone`, `SetTypeFlags` and `UpdateTypeFlags` to provide easy way to update `TypeCtx`
6. Introduce `NoCopy` struct to mark an object as no copy.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -420,6 +425,55 @@ type StatementContext struct {
 	}
 }
 
+// NewStmtCtx creates a new statement context
+func NewStmtCtx() *StatementContext {
```
Comment (YangKeao) [inline sessionctx/stmtctx/stmtctx.go:429]: I've tried to write a linter to make sure the `&StatementContext{}` is only constructed manually in `NewStmtCtx` or `NewStmtCtxWithTimeZone`. I'll submit that linter after this PR is merged 🍻

**Review by YangKeao — COMMENTED** [review]
Rest LGTM

**Review by YangKeao — APPROVED** [review]
No Comment

**Review by lance6716 — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**